pkg/oci/oci.go

```diff
@@ -478,16 +490,141 @@ func newAuthClient(repository string, authOptions *AuthOptions) (*oras_auth.Clie
+	// WARNING:
+	// https://sigstore.slack.com/archives/C0440BFT43H/p1712253122721879?thread_ts=1712238666.552719&cid=C0440BFT43H
+	// https://github.com/sigstore/cosign/pull/3622
+	parts := strings.Split(digest, ":")
+	if len(parts) != 2 {
+		return "", fmt.Errorf("wrong digest, expected two parts, got %d", len(parts))
+	}
+
+	return fmt.Sprintf("%s-%s.sig", parts[0], parts[1]), nil
+}
+
+func getSigningInformation(ctx context.Context, image string, authOpts *AuthOptions) (string, string, error) {
+	imageStore, err := getLocalOciStore()
+	if err != nil {
+		return "", "", fmt.Errorf("getting local oci store: %w", err)
+	}
+
+	imageRef, err := normalizeImageName(image)
+	if err != nil {
+		return "", "", fmt.Errorf("normalizing image name: %w", err)
+	}
+
+	desc, err := imageStore.Resolve(ctx, imageRef.String())
+	if err != nil {
+		return "", "", fmt.Errorf("resolving image %q: %w", image, err)
+	}
+
+	imageDigest := desc.Digest.String()
+	signatureTag, err := craftSignatureTag(imageDigest)
+	if err != nil {
+		return "", "", fmt.Errorf("crafting signature tag: %w", err)
+	}
+
+	repo, err := newRepository(imageRef, authOpts)
+	if err != nil {
+		return "", "", fmt.Errorf("creating repository: %w", err)
+	}
+
+	_, signatureBytes, err := oras.FetchBytes(ctx, repo, signatureTag, oras.DefaultFetchBytesOptions)
+	if err != nil {
+		return "", "", fmt.Errorf("getting signature bytes: %w", err)
+	}
+
+	signatureManifest := &ocispec.Manifest{}
+	err = json.Unmarshal(signatureBytes, signatureManifest)
+	if err != nil {
+		return "", "", fmt.Errorf("decoding signature manifest: %w", err)
+	}
+
+	layers := signatureManifest.Layers
+	expectedLen := 1
+	layersLen := len(layers)
+	if layersLen != expectedLen {
+		return "", "", fmt.Errorf("wrong number of signature manifest layers: expected %d, got %d", expectedLen, layersLen)
+	}
+
+	layer := layers[0]
+	expectedMediaType := "application/vnd.dev.cosign.simplesigning.v1+json"
+	if layer.MediaType != expectedMediaType {
+		return "", "", fmt.Errorf("wrong layer media type: expected %s, got %s", expectedMediaType, layer.MediaType)
+	}
+
+	signature, ok := layer.Annotations["dev.cosignproject.cosign/signature"]
+	if !ok {
+		return "", "", fmt.Errorf("no signature in layer")
+	}
+
+	payloadTag := layer.Digest.String()
+	// The payload is stored as a blob, so we fetch bytes from the blob store and
+	// not the manifest one.
+	_, payloadBytes, err := oras.FetchBytes(ctx, repo.Blobs(), payloadTag, oras.DefaultFetchBytesOptions)
+	if err != nil {
+		return "", "", fmt.Errorf("getting payload bytes: %w", err)
+	}
+
+	img := &payload.SimpleContainerImage{}
+	err = json.Unmarshal(payloadBytes, img)
+	if err != nil {
+		return "", "", fmt.Errorf("decoding payload: %w", err)
+	}
+
+	if img.Critical.Image.DockerManifestDigest != imageDigest{
+		return "", "", fmt.Errorf("payload digest does not correspond to image: expected %s, got %s", imageDigest, img.Critical.Image.DockerManifestDigest)
+	}
```

Comment:
According to my understanding from https://github.com/containers/image/blob/main/docs/containers-signature.5.md#the-cryptographic-signature, this logic should only be executed after `VerifySignature` returns nil:

> Consumers of container signatures SHOULD verify the cryptographic signature against one or more trusted public keys (e.g. defined in a [policy.json signature verification policy file](https://github.com/containers/image/blob/main/docs/containers-policy.json.5.md)) before parsing or processing the JSON payload in any way, in particular they SHOULD stop processing the container signature if the cryptographic signature verification fails, without even starting to process the JSON payload.